x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java

```diff
@@ -53,7 +69,18 @@ public JoinConfig config() {
 
     @Override
     protected NodeInfo<Join> info() {
-        return NodeInfo.create(this, Join::new, left(), right(), config);
+        // Do not just add the JoinConfig as a whole - this would prevent correctly registering the
+        // expressions and references.
+        return NodeInfo.create(
+            this,
+            Join::new,
+            left(),
+            right(),
+            config.type(),
+            config.matchFields(),
+            config.leftFields(),
+            config.rightFields()
```

Comment:
The crux: having the join config's parts actually in the node info, rather than the join config as a whole.

This makes it so that the fields the Join relies on are actually picked up when we run `expressions` - and therefore also when we run `references`.